applications/convection_diffusion_application/custom_response_functions/local_temperature_average_response_function.h

```diff
@@ -0,0 +1,246 @@
+                           const Matrix& rResidualGradient,
+                           Vector& rResponseGradient,
+                           const ProcessInfo& rProcessInfo) override
+    {
+        noalias(rResponseGradient) = ZeroVector(rResidualGradient.size1());
+    }
+
+    void CalculateFirstDerivativesGradient(const Element& rAdjointElement,
+                                           const Matrix& rResidualGradient,
+                                           Vector& rResponseGradient,
+                                           const ProcessInfo& rProcessInfo) override
+    {
+        ComputePointTemperatureSensitivityContribution(rResidualGradient, rAdjointElement.GetGeometry().Points(),rResponseGradient);
+    }
+
+    void CalculateSecondDerivativesGradient(const Element& rAdjointElement,
+                                            const Matrix& rResidualGradient,
+                                            Vector& rResponseGradient,
+                                            const ProcessInfo& rProcessInfo) override
+    {
+        ComputePointTemperatureSensitivityContribution(rResidualGradient, rAdjointElement.GetGeometry().Points(),rResponseGradient);
+    }
+
+    void CalculatePartialSensitivity(Element& rAdjointElement,
+                                     const Variable<array_1d<double, 3>>& rVariable,
+                                     const Matrix& rSensitivityMatrix,
+                                     Vector& rSensitivityGradient,
+                                     const ProcessInfo& rProcessInfo) override
+    {
+        if (rSensitivityGradient.size() != rSensitivityMatrix.size1())
+            rSensitivityGradient.resize(rSensitivityMatrix.size1(), false);
+        noalias(rSensitivityGradient) = ZeroVector(rSensitivityMatrix.size1());
+    }
+
+    void CalculatePartialSensitivity(Condition& rAdjointElement,
+                                     const Variable<array_1d<double, 3>>& rVariable,
+                                     const Matrix& rSensitivityMatrix,
+                                     Vector& rSensitivityGradient,
+                                     const ProcessInfo& rProcessInfo) override
+    {
+        if (rSensitivityGradient.size() != rSensitivityMatrix.size1())
+            rSensitivityGradient.resize(rSensitivityMatrix.size1(), false);
+        noalias(rSensitivityGradient) = ZeroVector(rSensitivityMatrix.size1());
+    }
+
+    double CalculateValue(ModelPart& rModelPart) override
+    {
+        KRATOS_TRY;
+        KRATOS_ERROR
+            << "PointTemperature::CalculateValue(ModelPart& rModelPart) is not implemented!!!\n";
+        KRATOS_CATCH("");
+    }
+
+    ///@}
+
+protected:
+    ///@name Protected member Variables
+    ///@{
+
+    ///@}
+    ///@name Protected Operators
+    ///@{
+
+    ///@}
+    ///@name Protected Operations
+    ///@{
+
+    ///@}
+
+private:
+    ///@name Member Variables
+    ///@{
+
+    ModelPart& mrModelPart;
+	int mNumNodes;
```

Comment:
In-class initialization is safer
```suggestion
	int mNumNodes = 0;
```